kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ProduceRequest.java

```diff
@@ -0,0 +1,305 @@
+import com.google.protobuf.ByteString;
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+import io.confluent.kafka.serializers.subject.RecordNameStrategy;
+import io.confluent.kafka.serializers.subject.TopicNameStrategy;
+import io.confluent.kafka.serializers.subject.TopicRecordNameStrategy;
+import io.confluent.kafka.serializers.subject.strategy.SubjectNameStrategy;
+import io.confluent.kafkarest.entities.EmbeddedFormat;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+@AutoValue
+public abstract class ProduceRequest {
+
+  ProduceRequest() {
+  }
+
+  @JsonProperty("partition_id")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<Integer> getPartitionId();
+
+  @JsonProperty("headers")
+  @JsonInclude(Include.NON_EMPTY)
+  public abstract ImmutableList<ProduceRequestHeader> getHeaders();
+
+  @JsonProperty("key")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<ProduceRequestData> getKey();
+
+  @JsonProperty("value")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<ProduceRequestData> getValue();
+
+  @JsonProperty("timestamp")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<Instant> getTimestamp();
+
+  @JsonCreator
+  static ProduceRequest fromJson(
+      @JsonProperty("partition_id") @Nullable Integer partitionId,
+      @JsonProperty("headers") @Nullable List<ProduceRequestHeader> headers,
+      @JsonProperty("key") @Nullable ProduceRequestData key,
+      @JsonProperty("value") @Nullable ProduceRequestData value,
+      @JsonProperty("timestamp") @Nullable Instant timestamp) {
+    return builder()
+        .setPartitionId(partitionId)
+        .setHeaders(headers != null ? headers : ImmutableList.of())
+        .setKey(key)
+        .setValue(value)
+        .setTimestamp(timestamp)
+        .build();
+  }
+
+  public static Builder builder() {
+    return new AutoValue_ProduceRequest.Builder().setHeaders(emptyList());
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder setPartitionId(@Nullable Integer partitionId);
+
+    public abstract Builder setHeaders(List<ProduceRequestHeader> headers);
+
+    public abstract Builder setKey(@Nullable ProduceRequestData key);
+
+    public abstract Builder setValue(@Nullable ProduceRequestData value);
+
+    public abstract Builder setTimestamp(@Nullable Instant timestamp);
+
+    public abstract ProduceRequest build();
+  }
+
+  @AutoValue
+  public abstract static class ProduceRequestHeader {
+
+    ProduceRequestHeader() {
+    }
+
+    @JsonProperty("name")
+    public abstract String getName();
+
+    @JsonIgnore
+    public abstract Optional<ByteString> getValue();
+
+    @JsonProperty("value")
+    @JsonInclude(Include.NON_ABSENT)
+    final Optional<BinaryNode> getSerializedValue() {
+      return getValue().map(value -> BinaryNode.valueOf(value.toByteArray()));
```

Comment:
header values are byte arrays, but I think a lot of the time people will use strings, and it would be convenient to allow for this with a type property. i guess fine to leave as possible future enhancement, though maybe it affects the types you use in this class. not that they couldn't be changed in the future as well.

Reply:
I'm using bytes here to allow everything that the Java API allows, but ideally this would be ASCII encoded string, which is what is used for HTTP headers. We can extend/add an option later for string header values. I will skip it for now.